rgw: metadata refactoring

cbodley opened the pull request:
rebase of https://github.com/ceph/ceph/pull/28679 with some additional cleanup

cbodley commented:
cc @yehudasa

cbodley commented:
some radosgw-admin failures, and lots of valgrind issues to go through: http://pulpito.ceph.com/cbodley-2019-07-26_18:05:13-rgw-wip-cbodley-testing-distro-basic-smithi/

cbodley commented:
from qa/tasks/radosgw_admin.py, the creation of user2=fud fails during `# create a second user to link the bucket to`, because it finds an access_key2 object pointing at user1=foo. this key is added to user1 in `# TESTCASE 'add-keys','key','create','w/valid info','succeeds'` and removed in `# TESTCASE 'rm-key','key','rm','newly added key','succeeds, key is removed'`, but it looks like the rados object doesn't get deleted

cbodley commented on src/rgw/rgw_user.cc:
```diff
@@ -1208,7 +875,6 @@ int RGWAccessKeyPool::execute_remove(RGWUserAdminOpState& op_state, std::string
     return -ERR_INVALID_ACCESS_KEY;
   }
 
-  rgw_remove_key_index(store, kiter->second);
```
these calls to rgw_remove_key_index() were all removed. it looks like the responsibility to delete them was moved into `store_user_info()`, but its PutOperation::remove_old_indexes() didn't handle the access key objects. i added a commit to do this; does that looks right @yehudasa?

yehudasa replied:
@cbodley yup!

cbodley commented:
the valgrind leak was because RGWSI_MetaBackend_Handler::Op doesn't have a virtual destructor, so the derived Op_ManagedCtx destructor isn't called to delete the allocated Context. since Op doesn't have any other virtual methods, i decided against making its destructor virtual. instead, i pushed a commit that uses the public Op_ManagedCtx directly

cbodley commented:
next failure is in [radosgw_admin_rest](http://qa-proxy.ceph.com/teuthology/cbodley-2019-07-30_12:00:27-rgw-wip-cbodley-testing-distro-basic-smithi/4165244/teuthology.log) on test case `# try creating an object with the first user which should cause an error` after 'bucket link' with another user:
```
   File "qa/tasks/radosgw_admin_rest.py", line 520, in task
    assert denied
```

cbodley commented on src/rgw/rgw_bucket.cc:
```diff
@@ -876,42 +675,47 @@ int RGWBucket::link(RGWBucketAdminOpState& op_state, std::string *err_msg)
+
+    // ...and encode the acl
+    attrs[RGW_ATTR_ACL].clear();
+    policy.encode(attrs[RGW_ATTR_ACL]);
```
this was missing the call to clear(), so re-encoding the policy was appending to the old one and not taking effect

yehudasa replied:
@cbodley looks good, but maybe can do it like this:
`auto& attr = atts[RGW_ATTR_ACL];
attr.clear();
policy.encode(attr);
`

to avoid a double lookup?

cbodley commented:
http://pulpito.ceph.com/cbodley-2019-08-01_13:07:49-rgw-wip-cbodley-testing-distro-basic-smithi/ looks like a pass. i'll start rebasing this on top of https://github.com/ceph/ceph/pull/28813 so we can merge both

cbodley commented:
i squashed everything into a single commit (because individual commits did not compile on their own), and rebased on top of https://github.com/ceph/ceph/pull/29540

yehudasa commented:
> i squashed everything into a single commit (because individual commits did not compile on their own), and rebased on top of #29540

@cbodley Then we'll lose all history that I worked really hard to retain when rebasing it the first time. How about doing a git merge instead of rebase?

mattbenjamin commented:
@yehudasa how much of the history is history just of the evolution of metadata refactor in your checkout?

yehudasa commented:
@mattbenjamin I can't really say. History can have value, even if not immediately clear. Squashing everything and applying a single commit is technically the same as running git merge. Is there a reason to lose all the history?

cbodley commented on src/rgw/rgw_user.h:
```diff
@@ -782,8 +773,192 @@ class RGWUserAdminOp_Caps
+    RGWObjVersionTracker *objv_tracker{nullptr};
+    ceph::real_time mtime;
+    bool exclusive{false};
+    map<string, bufferlist> *attrs;
```
uninitialized pointer crashing 'user rename' test, will add a commit for this with next rebase

cbodley commented:
mostly passed in http://pulpito.ceph.com/cbodley-2019-08-13_12:16:43-rgw-wip-cbodley-testing-distro-basic-smithi/

some `403 Forbidden` failures from radosgw-admin tests because i missed a couple pieces of 'user rename' in the last rebase. i added those back and pushed the branch after verifying the fix locally. i'd like to merge this as is to unblock other work; any objections @yehudasa @dang?

yehudasa commented:
@cbodley no objections

votdev commented:
Please run the Dashboard QA test suites before next time to ensure RGW changes do not break the Dashboard or inform the Dashboard team and open a tracker issue with the changes that needs to be adapted in the Dashboard.

cbodley commented:
@votdev i don't think it's reasonable to run every rgw pr through an extra teuthology suite. i'd prefer to move the dashboard's functional test coverage of rgw admin apis into qa/tasks/radosgw_admin_rest.py which runs as part of the rgw suite

cbodley commented on src/rgw/rgw_bucket.cc:
```diff
@@ -994,7 +649,6 @@ int RGWBucket::link(RGWBucketAdminOpState& op_state,
     return -EINVAL;
   }
   rgw_bucket old_bucket = bucket;
-  bucket.tenant = tenant;
```
oops, somehow this change snuck into the rebase
